File: solutions/medium/minstack.js

var MinStack = function() {
    this.stack = [];
};


/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    this.stack.push(val)
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    this.stack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length - 1]; 
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
   let temp = [];
   let min = this.stack[this.stack.length-1];
while(this.stack.length>0){
    min = Math.min(min,this.stack[this.stack.length-1]);

    temp.push(this.stack.pop());

}

while(temp.length>0){
    this.stack.push(temp.pop())
}
return min;

};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: solutions/JavaSolutions/MinStack.java
import java.util.Stack;

public class MinStack {
    private Stack<Integer> stack;

    public MinStack() {
this.stack = new Stack<>();
    }

    public void push(int val) {
stack.push(val);

    }

    public void pop() {
stack.pop();
    }

    public int top() {
      return stack.peek();
    }

    public int getMin() {

       Stack<Integer> temp = new Stack<>();
       int min = stack.peek();
while(!stack.isEmpty()){
    min = Math.min(min,stack.peek());
    temp.push(min);
}

while(!temp.isEmpty()){
    stack.push(temp.pop());
}
return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */